### ISSUE TITLE
Does NNEF supports object detection algorithms like Tiny YOLO?

### ISSUE THREAD

**Issue by tusharc7360** [KhronosGroup/NNEF-Tools#30]
I want to convert tiny-yolo-voc.cfg to NNEF format and  NNEF tool is supporting converter for caffe and tensorflow only. So, in that case I have converted tiny-yolo-voc.cfg to caffe prototxt file with the help of open source tool. Then the corresponding prototxt file given to converter tool by running the following command:
python export_nnef_description.py --graph ../out.prototxt --weights ../out.caffemodel 

The above step throws error:
[libprotobuf ERROR google/protobuf/text_format.cc:245] Error parsing text-format caffe.NetParameter: 412:18: Message type "caffe.LayerParameter" has no field named "region_param".
F0605 15:14:36.965162 24468 upgrade_proto.cpp:90] Check failed: ReadProtoFromTextFile(param_file, param) Failed to parse NetParameter file: ../out.prototxt
*** Check failure stack trace: ***
Aborted (core dumped)

I think the problem is with the last  layer "Region" as the NNEF specification does not talk about it.
"Region" layer is giving the bounding box information.

So , my question is that if NNEF supports "Region" layer. If not, then how can I incorporate "Region" layer into NNEF format.

Thanks

**Comment by tusharc7360** [comment]
I have tried to run the model in Caffe. It gives the same error.
Darknet framework was used to train the original model.

**Comment by gyenesvi** [comment]
Unfortunately, there is no Darknet support at the moment, and NNEF does not have region layer support either. We will consider it in a next release.

**Comment by tusharc7360** [comment]
Okay. Thanks for the information @gyenesvi. 
Next release will be the final version of 1.0 right?
When we can expect the next release?

**Comment by gyenesvi** [comment]
Yes, the next release is going to be 1.0 final, and that is almost finalized, coming soon. Sorry, I meant the next release after 1.0.

**Comment by tusharc7360** [comment]
Okay, Thanks a lot @gyenesvi.

**Comment by tusharc7360** [comment]
HI @gyenesvi , I have a question, if NNEF is not supporting region layer at the moment, then how can I build my NNEF format for Tiny YOLO which have region layer.

**Comment by gyenesvi** [comment]
It seems to me that the region layer performs the generation of the actual bounding boxes. I have found some TensorFlow and PyTorch implementations on github:

https://github.com/mystic123/tensorflow-yolo-v3
https://github.com/ne7ermore/yolo-v3

In these implementations, the detections are assembled as a sequence of more primitive operations, which are (mostly) supported in NNEF. At a first glance, the only problematic operation seems to be the `tile`/`repeat` operation, which we are also considering to add to NNEF. If that happens, it should be possible to export the TensorFlow version into NNEF.

**Comment by tusharc7360** [comment]
Hi @gyenesvi, so what you want to say  is that the repeat Tensors operations like tf.tile is the problematic operation.  Also I have gone through the tf2nnef.py script, there DefaultExporters does not seem to have t.ftile operation. I Understood right @gyenesvi ?

**Comment by gyenesvi** [comment]
Yes, that's right. Furthermore, the `tf.tile` operation currently does not have a counterpart in NNEF, that's why it's not exported.

**Comment by tusharc7360** [comment]
Hi @gyenesvi , one more basic question, how can I integrate NNEF parser to the C code where I want to get the network definition through any nnef file, build the network model and then execute the network model. Indirectly, I am asking how to use the NNEF tool.

**Comment by gyenesvi** [comment]
You can integrate the C++ code in this repo as you wish. It will only work with a C++ compiler. The actual NNEF parser code is in the header files, so you have to include those to your project. The main.cpp gives you an example of how to invoke the parser (much of main.cpp is just argument parsing and printing back the resulting graph). You have to include "comp/comp_parser.h" only if you need full NNEF functionality.

Basically you write your own callback class, invoke the parser with an input stream and the callback, and then handle each operation in the callback. That's where you can build your network model and execute it afterwards.

Please, if you have further questions, ask them in separate issues, so that the conversation threads don't get mixed up.

**Comment by tusharc7360** [comment]
Thanks for information @gyenesvi.
Will raise the issue in separate thread.
Closing the issue.
